gen/field.go

```diff
@@ -81,57 +86,95 @@ func (f fieldGroupGenerator) DefineStruct(g Generator) error {
 }
 
 func (f fieldGroupGenerator) ToWire(g Generator) error {
-	// TODO(abg): Default values
 	return g.DeclareFromTemplate(
 		`
 		<$wire := import "github.com/thriftrw/thriftrw-go/wire">
 
 		<$v := newVar "v">
-		func (<$v> *<.Name>) ToWire() <$wire>.Value {
+		func (<$v> *<.Name>) ToWire() (<$wire>.Value, error) {
     		<$fields := newVar "fields">
     		<$i := newVar "i">
-			// TODO check if required fields that are reference types are nil
+			<$wVal := newVar "w">
 
-			var <$fields> [<len .Fields>]<$wire>.Field
-			<$i> := 0
+			<if and .IsUnion (len .Fields)>
+				<$fmt := import "fmt">
+				<$count := newVar "count">
+				<$count> := 0
+				<range .Fields>
+					if <$v>.<goCase .Name> != nil { <$count>++ }
+				<end>
+				<if .AllowEmptyUnion>
+					if <$count> > 1 {
+						return <$wire>.Value{}, <$fmt>.Errorf(
+							"<.Name> should receive at most one field value: received %v values", <$count>)
+					}
+				<else>
+					if <$count> != 1 {
+						return <$wire>.Value{}, <$fmt>.Errorf(
+							"<.Name> should receive exactly one field value: received %v values", <$count>)
+					}
+				<end>
+			<end>
+
+			var (
+					<$fields> [<len .Fields>]<$wire>.Field
+					<$i> int = 0
+				<if len .Fields>
+					<$wVal> <$wire>.Value
+					err error
+				<end>
+			)
 
+			<$structName := .Name>
 			<range .Fields>
 				<$f := printf "%s.%s" $v (goCase .Name)>
 				<if .Required>
-					<$wVal := toWire .Type $f>
-					<$fields>[<$i>] = <$wire>.Field{ID: <.ID>, Value: <$wVal>}
-					<$i>++
-				<else>
-					<if .Default>
+					<if not (isPrimitiveType .Type)>
 						if <$f> == nil {
-							<$f> = <constantValuePtr .Default .Type>
+							return <$wVal>, <import "errors">.New(
+								"field <goCase .Name> of <$structName> is required")
+						}
+					<end>
+						<$wVal>, err = <toWire .Type $f>
+						if err != nil {
+							return <$wVal>, err
 						}
 						<$fields>[<$i>] = <$wire>.Field{
 							ID: <.ID>,
-							Value: <toWirePtr .Type $f>,
+							Value: <$wVal>,
 						}
 						<$i>++
+				<else>
+					<if .Default>
+						if <$f> == nil {
+							<$f> = <constantValuePtr .Default .Type>
+						}
 					<else>
 						if <$f> != nil {
+					<end>
+							<$wVal>, err = <toWirePtr .Type $f>
+							if err != nil {
+								return <$wVal>, err
+							}
 							<$fields>[<$i>] = <$wire>.Field{
 								ID: <.ID>,
-								Value: <toWirePtr .Type $f>,
+								Value: <$wVal>,
 							}
 							<$i>++
+					<if not .Default>
```

Comment:
this is kind of confusing. what might be clearer is even in the `if .Default` case,  to just open a new brace, so that you always need to close regardless.